exonum_derive/src/pb_convert.rs

```diff
@@ -0,0 +1,165 @@
+        if nv.ident == "protobuf_convert" {
+            match nv.lit {
+                Lit::Str(path) => Some(path.parse::<Path>().unwrap()),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    });
+
+    struct_path.expect("protobuf_convert attribute is not set properly.")
+}
+
+fn gen_protobuf_convert_from_pb(field_names: &[Ident]) -> impl quote::ToTokens {
+    let getters = field_names
+        .iter()
+        .map(|i| Ident::new(&format!("get_{}", i), Span::call_site()));
+    let our_struct_names = field_names.to_vec();
+
+    quote! {
+        fn from_pb(pb: Self::ProtoStruct) -> std::result::Result<Self, ()> {
+          Ok(Self {
+           #( #our_struct_names: ProtobufConvert::from_pb(pb.#getters().to_owned())?, )*
+          })
+        }
+    }
+}
+
+fn gen_protobuf_convert_to_pb(field_names: &[Ident]) -> impl quote::ToTokens {
+    let setters = field_names
+        .iter()
+        .map(|i| Ident::new(&format!("set_{}", i), Span::call_site()));
+    let our_struct_names = field_names.to_vec();
+
+    quote! {
+        fn to_pb(&self) -> Self::ProtoStruct {
+            let mut msg = Self::ProtoStruct::new();
+            #( msg.#setters(ProtobufConvert::to_pb(&self.#our_struct_names).into()); )*
+            msg
+        }
+    }
+}
+
+fn gen_to_protobuf_impl(
+    name: &Ident,
+    pb_name: &Path,
+    field_names: &[Ident],
+) -> impl quote::ToTokens {
+    let to_pb_fn = gen_protobuf_convert_to_pb(field_names);
+    let from_pb_fn = gen_protobuf_convert_from_pb(field_names);
+
+    quote! {
+        impl ProtobufConvert for #name {
+            type ProtoStruct = #pb_name;
+
+            #to_pb_fn
+            #from_pb_fn
+        }
+    }
+}
+
+fn gen_binary_form_impl(name: &Ident, cr: &quote::ToTokens) -> impl quote::ToTokens {
+    quote! {
+        impl #cr::messages::BinaryForm for #name {
+
+            fn encode(&self) -> std::result::Result<Vec<u8>, _EncodingError> {
+                self.to_pb().write_to_bytes().map_err(|e| _EncodingError::Other(Box::new(e)))
+            }
+
+            fn decode(buffer: &[u8]) -> std::result::Result<Self, _EncodingError> {
+                let mut pb = <Self as ProtobufConvert>::ProtoStruct::new();
+                pb.merge_from_bytes(buffer).map_err(|e| _EncodingError::Other(Box::new(e)))?;
+                Self::from_pb(pb).map_err(|_| "Conversion from protobuf error.".into())
+            }
+        }
+    }
+}
+
+fn gen_storage_traits_impl(name: &Ident, cr: &quote::ToTokens) -> impl quote::ToTokens {
+    quote! {
+        impl #cr::crypto::CryptoHash for #name {
+            fn hash(&self) -> #cr::crypto::Hash {
+                let v = self.to_pb().write_to_bytes().unwrap();
+                #cr::crypto::hash(&v)
+            }
+        }
+
+        impl #cr::storage::StorageValue for #name {
+            fn into_bytes(self) -> Vec<u8> {
+                self.to_pb().write_to_bytes().unwrap()
```

Comment:
Nit: I'd switch `unwrap()` to `expect()` with more informative messages here and in `from_bytes`. Also, it could make sense to add a comment explaining why unwrapping is fine here (because we deal with trusted data source).